### PR TITLE
F-26: new vm template attributes

### LINKED ISSUE #26: Add template and VM schemas
As proposed in this comment: https://github.com/OpenNebula/terraform-provider-opennebula/pull/24#discussion_r377229103

We could add some new schemas parts for VM and template resources.

Draft proposal:

We could add the field `description` on both (but also on images... etc)

Add to Template resource:
- `hypervisor`
- `sched_requirements`
- `user_inputs`

EDIT: an user asked for an other [placement section attribute](https://docs.opennebula.io/5.12/operation/references/template.html#placement-section):
- `shed_ds_requirements`

Feel free to drop a comment to update this list

### PR DESCRIPTION
**Pull request by treywelsh**
adds hypervisor, description, user_inputs, sched_requirements sections to VM and Template resources schemas.

There is no acceptance test for these attributes for now

### CODE REVIEW

File: opennebula/resource_opennebula_image.go [hunk 1]
```diff
@@ -580,6 +580,18 @@ func resourceOpennebulaImageUpdate(d *schema.ResourceData, meta interface{}) err
 		log.Printf("[INFO] Successfully updated Image Type %s\n", image.Name)
 	}
 
+	if d.HasChange("description") {
```
Comment (jaypif) [inline opennebula/resource_opennebula_image.go]: why not doing update once if there are several updates like it is already the case for templates or VMs ?

File: opennebula/shared_schemas.go [hunk 2]
```diff
@@ -238,6 +238,22 @@ func tagsSchema() *schema.Schema {
 	}
 }
 
+func schedReqSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:        schema.TypeString,
+		Optional:    true,
+		Description: "Boolean expression that rules out provisioning hosts from list of machines suitable to run this VM.",
```
Comment (jaypif) [inline opennebula/shared_schemas.go]: Are you sure the description is correct ?
check y suggestion in the doc part.

File: website/docs/r/virtual_machine.html.markdown [hunk 3]
```diff
@@ -87,6 +91,7 @@ The following arguments are supported:
 * `nic` - (Optional) Can be specified multiple times to attach several NICs. See [Nic parameters](#nic-parameters) below for details.
 * `vmgroup` - (Optional) See [VM group parameters](#vm-group-parameters) below for details. Changing this argument triggers a new resource.
 * `group` - (Optional) Name of the group which owns the virtual machine. Defaults to the caller primary group.
+* `sched_requirements` - (Optional) A boolean expression to select the hosts that could run the VM.
```
Comment (jaypif) [inline website/docs/r/virtual_machine.html.markdown]: I do prefer something like: Scheduling requirements to deploy the resource following specific rule.

**Comment by treywelsh** [comment]
By the way, I added `sched_ds_requirements` attribute from placement section

**Comment by CRASH-Tech** [comment]
Please, appruve this PR

File: website/docs/r/template.html.markdown [hunk 4]
```diff
@@ -82,6 +90,9 @@ The following arguments are supported:
 * `nic` - (Optional) Can be specified multiple times to attach several NICs. See [Nic parameters](#nic-parameters) below for details.
 * `raw` - (Optional) Allow to pass hypervisor level tuning content. See [Raw parameters](#raw-parameters) below for details.
 * `vmgroup` - (Optional) See [VM group parameters](#vm-group-parameters) below for details. Changing this argument triggers a new resource.
+* `user_inputs` - (Optional) Ask the user instantiating the template to define the values described.
+* `sched_requirements` - (Optional) Scheduling requirements to deploy the resource following specific rule
+* `sched_ds_requirements` - (Optional) Storage requirements to deploy the resource following specific rule.
```
Comment (jaypif) [inline website/docs/r/template.html.markdown]: It is "storage placement requirements" to be correct

File: opennebula/shared_schemas.go [hunk 5]
```diff
@@ -238,6 +238,30 @@ func tagsSchema() *schema.Schema {
...
+		Optional:    true,
+		Description: "Scheduling requirements to deploy the resource following specific rule",
+	}
+}
+
+func schedDSReqSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:        schema.TypeString,
+		Optional:    true,
+		Description: "Storage requirements to deploy the resource following specific rule",
```
Comment (jaypif) [inline opennebula/shared_schemas.go]: It is "storage placement requirements" to be correct

**Review by jaypif — CHANGES_REQUESTED** [review]
Add Unit test on these new attributes

**Comment by treywelsh** [comment]
It seems that the update behaviour may be improved in case of key deletion.
For now in this PR, in order to delete an element `key="value1"` I send a template with the empty value: `key=""`
Then, I call update method with merge behaviour (other behavior available is replace)

However, implement this behaviour will complexify the code: I would need to collect the template of the VM, delete the element, then call update with replace behavior.

**Comment by treywelsh** [comment]
I modified the update behavior to allow to delete the templates attributes from the VM and Template ressources.

**Review by jaypif — APPROVED** [review]

**Pull request merged**